### PR TITLE
WIP: Fixing CORS error in genericapiserver

### PR DESCRIPTION
**Pull request by nikhiljindal**
Ref https://github.com/kubernetes/kubernetes/issues/24086

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/27386)

### CODE REVIEW

**Comment by nikhiljindal** [comment]
I think I have a fix but I cant get the test to pass, which obviously means that my fix is not correct.
Am still debugging, but wanted to put this out so that maybe someone else spots the problem.
cc @Pendoragon @lavalamp 


When I run the test, I see that the CORS handler is registered. But it is not called when I send the request.
I get a 405 method not allowed response from go-restful when I call OPTIONS on /apis. This means the ServerMux handler is being called before my CORS handler, but I cant figure out why.

**Comment by smarterclayton** [comment]
Since this was listed as a regression, do we need to commit to this for 1.4?

File: pkg/genericapiserver/genericapiserver_test.go [hunk 1]
```diff
@@ -311,6 +313,29 @@ func TestDiscoveryAtAPIS(t *testing.T) {
 
 	assert.Equal(0, len(groupList.Groups))
 }
+*/
+
+func TestCORSRequest(t *testing.T) {
+	master, etcdserver, _, _ := newMaster(t)
+	defer etcdserver.Terminate(t)
+
+	server := httptest.NewServer(master.HandlerContainer.ServeMux)
```
Comment (Pendoragon) [inline pkg/genericapiserver/genericapiserver_test.go]: I think this should be `server := httptest.NewServer(master.Handler)` or `server := httptest.NewServer(master.InsecureHandler)`

File: pkg/genericapiserver/genericapiserver_test.go [hunk 2]
```diff
@@ -311,6 +313,29 @@ func TestDiscoveryAtAPIS(t *testing.T) {
...
+	req, err := http.NewRequest("OPTIONS", server.URL+"/apis", nil)
+	//	req, err := http.NewRequest("OPTIONS", server.URL, nil)
+	//	req, err := http.NewRequest("OPTIONS", server.URL+"/api/v1/namespaces/ns1/pods", nil)
+	req.Header.Add("Origin", "http://not-my-domain.com")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
```
Comment (Pendoragon) [inline pkg/genericapiserver/genericapiserver_test.go]: In case of an OPTIONS request, we are expecting `http.StatusNoContent`

**Pull request closed without merging**